tests/integration/test_partitioning_key.py

```diff
@@ -722,6 +723,25 @@
             (CAST('2023-01-01 11:55:59.999999' AS TIMESTAMP), CAST('2023-01-01' AS DATE), 'some data');
             """,
         ),
+        # Test that special characters are URL-encoded
+        (
+            [PartitionField(source_id=15, field_id=1001, transform=IdentityTransform(), name="special#string#field")],
+            ["special string"],
+            Record(**{"special#string#field": "special string"}),  # type: ignore
+            "special%23string%23field=special%20string",
+            f"""CREATE TABLE {identifier} (
+                `special#string#field` string
+            )
+            USING iceberg
+            PARTITIONED BY (
+                identity(`special#string#field`)
+            )
+            """,
+            f"""INSERT INTO {identifier}
+            VALUES
+            ('special string')
+            """,
+        ),
```

Comment:
Oops. I thought this test would work but it fails, and I'm not sure why yet.

It fails

```py
assert expected_hive_partition_path_slice in spark_path_for_justification
```

with

```
E AssertionError: assert 'special%23string%23field=special%20string' in 's3://warehouse/default/test_table/data/special#string#field=special+string/00000-57-0756f620-2b2e-4ffa-97f6-625343525c9b-00001.parquet'
```

and it fails

```py
assert spark_partition_for_justification == expected_partition_record
```

with

```
E AssertionError: assert Record[specia...ecial string'] == Record[specia...ecial string']
E - Record[special#string#field='special string']
E + Record[special_x23string_x23field='special string']
```

Reply:
`special_x23string_x23field` is related to https://github.com/apache/iceberg-python/pull/590

Reply:
Thanks @kevinjqliu. And the first failure is because the Iceberg version on `spark` was before https://github.com/apache/iceberg/pull/10329, so it's not URL-encoded (I think).

Given this, I've disabled justification with a message similar to other tests here where behaviour differs.

Reply:
Do you think this is sufficient (given we're testing `PartitionKey`'s `to_path`, it felt natural but I'm unsure)?

If not, happy to be pointed to somewhere where I can add an integration test similar to the one shown in the issue. Thanks!

Reply:
https://github.com/apache/iceberg/commit/795fea9446723cf9c2dbc3cac07883cace8997c9
should be available starting in `1.6.0`. 

It looks like pyspark is using an older library version, can you add this change as see if the test pass? 
https://github.com/apache/iceberg-python/pull/1462

Reply:
> Do you think this is sufficient (given we're testing PartitionKey's to_path, it felt natural but I'm unsure)?

yea we're testing `partition_to_path`, maybe add a test in `tests/table/test_partitioning.py`, which is not part of the integration test suite.

The integration test is a nice to have though. lets see if upgrading the iceberg library helps

Reply:
Thanks for this suggestion - bumping made the path test fail because `quote` was being used instead of `quote_plus` for encoding (the Java implementation encodes spaces to `+` which `quote` doesn't do).

For consistency, I've made the change to match Java behaviour (but can revert that). A unit test sounds good (and integration for justification checks would be great).

Reply:
nice! thats why we like integration tests :) 
let me merge #1462 first, and then we can rebase

Reply:
But the `Record` comparison still fails because of the non-optional sanitisation-transformation described in https://github.com/apache/iceberg/issues/10120. And, as it stands, the provided `Record` param is used [to check `key.partition`](https://github.com/apache/iceberg-python/blob/main/tests/integration/test_partitioning_key.py#L748) so can't be changed because that should be unsanitised, IIUC.

Think some test rewiring might be required - maybe providing a separate record param that's by default the other one, just for this justification check, but I wonder if we're really just testing spark behaviour then.